### PR TITLE
feat(42684): Override JSDoc for re-exported types / enum / classes / interfaces

### LINKED ISSUE #42684: Override JSDoc for re-exported types / enum / classes / interfaces
# Suggestion

## 🔍 Search Terms

* Override JSDoc
* JSDoc for re-export

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Add the ability to define / override documentation(JSDoc) for re-exported types / enum / classes / interfaces, etc.

## 📃 Motivating Example

In the case when the library developer decided to change the name of the type / enum / class / interface, etc., it would be useful to have two copies of the declaration with different names and the old name would be marked as @deprecated.
This refactoring practice is adopted for class methods, when a new method is created and the body of the old method is transferred into it, and a call to the new method is placed in the old method. The old method is then marked as @deprecated so that the library user has time to replace the name of the called method.

## 💻 Use Cases

An example of overriding JSDoc to mark a enum / interface as deprecated

```ts
// file: core/newFeature.ts
/** Some enum description **/
export enum NewFeatureEnum {
    // ...
}
/** Some interface description **/
export enum INewFeature<T extends NewFeatureEnum> {
    // ...
}

// file: index.ts
export {
  NewFeatureEnum,
  /** @deprecated use NewFeatureEnum */
  NewFeatureEnum as OldFeatureEnum,
  INewFeature,
  /** @deprecated use INewFeature */
  INewFeature as IOldFeature,
  // ...
} from 'core/newFeature';
```
At the moment, the JSDoc comment added to the re-export construct will not affect the final result in any way

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #42684

### CODE REVIEW

File: tests/cases/fourslash/jsdocDeprecated_suggestion15.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/// <reference path="fourslash.ts" />
+
+// @module: esnext
+// @filename: /a.ts
+////export const a = 1;
+////export const b = 1;
```
Comment (sandersn) [inline tests/cases/fourslash/jsdocDeprecated_suggestion15.ts:6]: does this apply to same-file export lists too, like `export {a, /** @deprecated */ b }`? If so, we should have a test for that as well.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -25066,9 +25066,13 @@ namespace ts {
...
 
             const localOrExportSymbol = getExportSymbolOfValueSymbolIfExported(symbol);
-            const sourceSymbol = localOrExportSymbol.flags & SymbolFlags.Alias ? resolveAlias(localOrExportSymbol) : localOrExportSymbol;
-            if (sourceSymbol.declarations && getDeclarationNodeFlagsFromSymbol(sourceSymbol) & NodeFlags.Deprecated && isUncalledFunctionReference(node, sourceSymbol)) {
-                addDeprecatedSuggestion(node, sourceSymbol.declarations, node.escapedText as string);
+            if (localOrExportSymbol.flags & SymbolFlags.Alias) {
+                checkDeprecatedAliasedSymbol(localOrExportSymbol, node);
+            }
+            else {
+                if (localOrExportSymbol.declarations && getDeclarationNodeFlagsFromSymbol(localOrExportSymbol) & NodeFlags.Deprecated && isUncalledFunctionReference(node, localOrExportSymbol)) {
```
Comment (sandersn) [inline src/compiler/checker.ts:25073]: this line should be broken up

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -39802,8 +39806,35 @@ namespace ts {
                     }
                 }
 
-                if (isImportSpecifier(node) && target.declarations?.every(d => !!(getCombinedNodeFlags(d) & NodeFlags.Deprecated))) {
```
Comment (sandersn) [inline src/compiler/checker.ts:39805]: this doesn't call isUncalledFunctionReference before, but now checkDeprecatedAliasedSymbol will. Is that OK? Does isUncalledFunctionReference have a quick exit for non-functions?

**Review by sandersn — CHANGES_REQUESTED** [review]

File: src/compiler/checker.ts (hunk 3 above)
Comment (a-tarasyuk) [reply src/compiler/checker.ts:39805]: @sandersn Thanks for the review. Yep, `isUncalledFunctionReference` is required for use in `checkIdentifier`. This does not affect the `ImportSpecifier` check, but in this case, it is a useless condition. I've changed `checkDeprecatedAliasedSymbol` to return the target of the alias. Does that make sense?

File: tests/cases/fourslash/jsdocDeprecated_suggestion15.ts (hunk 1 above)
Comment (a-tarasyuk) [reply tests/cases/fourslash/jsdocDeprecated_suggestion15.ts:6]: I've added test.

File: src/compiler/checker.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/compiler/checker.ts:25073]: Moved to the new helper `isDeprecatedSymbol`

**Review by sandersn — APPROVED** [review]

**Pull request merged**